### PR TITLE
@orta=> Sync Settings Progress Bar & WiFi messaging

### LINKED ISSUE #123: Disable or hide Sync button upon tap
Should it become a spinner or grey out?

### PR DESCRIPTION
**Pull request by sarahscott**
- [x] Added sync progress bar
- [x] Clean up code
- [x] Tests

Fixes #123

### CODE REVIEW

**Comment by orta** [comment]
looking good so far

**Comment by ArtsyMobile** [comment]
&nbsp; | 1 Error
  ------------- | ------------ 
    :no_entry_sign: | No CHANGELOG changes made



  &nbsp; | 2 Warnings
  ------------- | ------------ 
    :warning: | PR is classed as Work in Progress
    :warning: | Needs testing on a Phone if change is non-trivial




<p align="right">
  Generated by 
    <a href="https://github.com/KrauseFx/danger">danger</a>
  on 
  <i>2016-01-07</i>
</p>

**Comment by ArtsyMobile** [comment]
&nbsp; | 1 Error
  ------------- | ------------ 
    :no_entry_sign: | No CHANGELOG changes made



  &nbsp; | 1 Warning
  ------------- | ------------ 
    :warning: | Needs testing on a Phone if change is non-trivial




<p align="right">
  Generated by 
    <a href="https://github.com/KrauseFx/danger">danger</a>
  on 
  <i>2016-01-07</i>
</p>

**Comment by ArtsyMobile** [comment]
:white_check_mark: | No errors found
  ------------- | ------------ 



  &nbsp; | 1 Warning
  ------------- | ------------ 
    :warning: | Needs testing on a Phone if change is non-trivial




<p align="right">
  Generated by 
    <a href="https://github.com/KrauseFx/danger">danger</a>
  on 
  <i>2016-01-08</i>
</p>

**Comment by orta** [comment]
Failing tests are legit:
```
Failing tests:
	-[ARLabSettingsSyncViewControllerSpec test_viewing_sync_records__looks_right_with_existing_sync_records_as_ipad]
	-[ARLabSettingsSyncViewControllerSpec test_responding_to_network_changes__looks_right_with_poor_network_quality_as_iphone]
	-[ARLabSettingsSyncViewControllerSpec test_responding_to_network_changes__looks_right_with_poor_network_quality_as_ipad]
** TEST FAILED **
```

File: ArtsyFolio Tests/View Controllers/ARLabSettingsSyncViewControllerTests.m [hunk 1]
```diff
@@ -21,24 +27,37 @@ @interface ARLabSettingsSyncViewController ()
...
     it(@"looks right with no previous syncs", ^{
-        [navController pushViewController:subject animated:NO];
         expect(navController).to.haveValidSnapshot ();
     });
     
-    it(@"looks right with existing sync records", ^{
+    /// If the test device context is changed, it must instantiate a new navController from the storyboard or it won't size properly
+    itHasSnapshotsForDevices(@"looks right with existing sync records", ^{
+        /// This bool tells FBSnapshotTestCase to use size classes; unfortunately, the navigation bar on iPhone looks kind of strange in the snapshots but the actual SyncSettingsViewController looks as it should
+        [Expecta setUsesDrawViewHierarchyInRect:YES];
```
Comment (orta) [inline ArtsyFolio Tests/View Controllers/ARLabSettingsSyncViewControllerTests.m]: interesting, we're looking into using size classes in eigen, any advice here?

File: ArtsyFolio Tests/View Controllers/ARLabSettingsSyncViewControllerTests.m [hunk 2]
```diff
@@ -46,9 +65,71 @@ @interface ARLabSettingsSyncViewController ()
...
+        subject.viewModel.networkQuality = ARNetworkQualityOffline;
+        [subject beginAppearanceTransition:YES animated:NO];
+        [subject updateSubviewsAnimated:NO];
+        expect(navController).to.haveValidSnapshot ();
+    });
+});
+
+describe(@"during a sync", ^{
+    beforeEach(^{
+        OCMockObject *mockViewModel = [OCMockObject partialMockForObject:subject.viewModel];
```
Comment (orta) [inline ArtsyFolio Tests/View Controllers/ARLabSettingsSyncViewControllerTests.m]: this is a pretty reasonable use case for `OCMockObject` :+1:

Comment (orta) [inline ArtsyFolio Tests/View Controllers/ARLabSettingsSyncViewControllerTests.m]: wait, how does this work? you're create the object within the scope of that before block - then it exits scope and still happens to keep working?

File: Classes/Controllers/Settings/ARLabSettingsSyncViewController.m [hunk 3]
```diff
@@ -75,16 +82,101 @@ - (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(N
...
         cell.textLabel.textColor = UIColor.artsyHeavyGrey;
     }
 }
 
+- (void)updateSubviews
+{
+    [self updateSubviewsAnimated:YES];
+}
+
+- (void)updateSubviewsAnimated:(BOOL)animated
```
Comment (orta) [inline Classes/Controllers/Settings/ARLabSettingsSyncViewController.m]: good job on the animated `BOOL`

File: Classes/Controllers/Settings/ARSyncStatusViewModel.h [hunk 4]
```diff
@@ -46,4 +56,18 @@ typedef NS_ENUM(NSInteger, ARSyncImageNotification) {
...
+
+- (NSString *)titleText;
+- (NSString *)subtitleText;
+- (BOOL)shouldShowSyncButton;
+- (NSString *)syncButtonTitle;
+- (UIColor *)syncButtonColor;
+- (CGFloat)syncActivityViewAlpha;
+
+- (ARSyncImageNotification)currentSyncImageNotification;
+
```
Comment (orta) [inline Classes/Controllers/Settings/ARSyncStatusViewModel.h]: great comments in this file up there ^

File: Classes/Controllers/Settings/ARSyncStatusViewModel.m [hunk 5]
```diff
@@ -17,40 +19,181 @@ @implementation ARSyncStatusViewModel
...
+#pragma mark dependency injection
+
+- (NSUserDefaults *)defaults
+{
+    return _defaults ?: [NSUserDefaults standardUserDefaults];
+}
+
+- (void)dealloc
+{
+    [self.qualityIndicator stopObservingNetworkQuality];
```
Comment (orta) [inline Classes/Controllers/Settings/ARSyncStatusViewModel.m]: went looking for this, good job :+1:

File: Classes/Controllers/Settings/ARSyncStatusViewModel.m [hunk 6]
```diff
@@ -17,40 +19,181 @@ @implementation ARSyncStatusViewModel
...
 - (void)startSync
 {
-    self.isSyncing = YES;
+    self.currentSyncPercentDone = 0;
     [self.sync sync];
 }
 
+- (BOOL)isActivelySyncing
+{
+    return self.networkQuality != ARNetworkQualityOffline && self.sync.isSyncing;
```
Comment (orta) [inline Classes/Controllers/Settings/ARSyncStatusViewModel.m]: might be good to add brackets around this, at a quick glance I can't tell whether it is:

```
return (self.networkQuality != ARNetworkQualityOffline) && self.sync.isSyncing
return self.networkQuality != (ARNetworkQualityOffline && self.sync.isSyncing)
```

File: docs/CHANGELOG.yml [hunk 7]
```diff
@@ -1,6 +1,7 @@
 upcoming:
```
Comment (orta) [inline docs/CHANGELOG.yml]: these WIFI images are _huge_ in px - I don't have a problem with this, but can you please run https://imageoptim.com against them to make sure they're tiny on device.

File: ArtsyFolio Tests/View Controllers/ARLabSettingsSyncViewControllerTests.m (hunk 1 above)
Comment (sarahscott) [inline ArtsyFolio Tests/View Controllers/ARLabSettingsSyncViewControllerTests.m]: size classes are magical unless you're using expecta + fbsnapshottestcase. i can't tell if the problem is with Expecta, FBSnapshots, or ARExpectaExtensions right now so i'm making it a separate investigation.

**Comment by sarahscott** [comment]
I ran imageoptim on the images - happy to use smaller ones too if need be. 

I'm not sure what's going one with the snapshots once I enable the usesDrawViewHierarchyInRect bool and there's not much documentation around it (https://github.com/facebook/ios-snapshot-test-case#features). I've decided to make that a separate issue so the new features here can become a testable beta. Basically, enabling that value lets autolayout work properly for iPhone snapshots but totally messes up the navigation bar and size of the resulting image.

**Comment by orta** [comment]
If you're happy, and it passes locally, I'm happy to merge this with it red ( after a rebase ) - so I can ensure everything is :+1:

**Comment by ArtsyMobile** [comment]
:white_check_mark: | No errors found
  ------------- | ------------ 



  &nbsp; | 1 Warning
  ------------- | ------------ 
    :warning: | Needs testing on a Phone if change is non-trivial




<p align="right">
  Generated by 
    <a href="https://github.com/KrauseFx/danger">danger</a>
  on 
  <i>2016-01-10</i>
</p>

**Comment by sarahscott** [comment]
yayyy it's :green_heart: !

**Comment by orta** [comment]
Awesome, :+1: :+1: :+1:

**Pull request merged**